### PR TITLE
ASR tutorial

### PR DESCRIPTION
**Pull request by okuchaiev**
Tutorial can be run entirely in G Colab

Signed-off-by: Oleksii Kuchaiev <okuchaiev@nvidia.com>

### CODE REVIEW

File: tutorials/asr/ASR_with_NeMo.ipynb [hunk 1]
```diff
@@ -0,0 +1,920 @@
...
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "jpYXX-GslR0E",
+        "colab_type": "text"
+      },
+      "source": [
+        "There we go! We've put together a full training pipeline for the model and trained it for 50 epochs. Let's start Tensorboard to see how training went. Note that WER stands for Word Error Rate and the lower it is, the better.\n",
```
Comment (redoctopus) [inline tutorials/asr/ASR_with_NeMo.ipynb:650]: New section here as well to distinguish that this is separate from building the model. Maybe "Evaluating the Training with Tensorboard"? Not sure.

File: tutorials/asr/ASR_with_NeMo.ipynb [hunk 2]
```diff
@@ -0,0 +1,920 @@
...
+        "You can definetely imporove this run by playng with hyperparametrers."
```
Comment (redoctopus) [inline tutorials/asr/ASR_with_NeMo.ipynb:651]: s/imporove/improve

File: tutorials/asr/ASR_with_NeMo.ipynb [hunk 3]
```diff
@@ -0,0 +1,920 @@
...
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "PSJeNwkTXQjA",
+        "colab_type": "text"
+      },
+      "source": [
+        "You can use your freshly trained model to transcribe audio in much the same way as you did with cloud model. (Though full-fledged model from cloud is, of course, better)"
```
Comment (redoctopus) [inline tutorials/asr/ASR_with_NeMo.ipynb:708]: Missing period at end of sentence

File: tutorials/asr/ASR_with_NeMo.ipynb [hunk 4]
```diff
@@ -0,0 +1,920 @@
...
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "PSJeNwkTXQjA",
+        "colab_type": "text"
+      },
+      "source": [
```
Comment (redoctopus) [inline tutorials/asr/ASR_with_NeMo.ipynb:707]: Add section about batch inference w/ WER evaluation?
I think it'd be useful to have a section (applicable to both the from-scratch and pretrained models) detailing how to do both individual sample transcription and batch inference.

Header: Inference
- Subheader: Individual Sample Inference (e.g. what is shown here)
- Subheader: Batch Inference (e.g. how to pass in a validation set and get WER)

File: tutorials/asr/ASR_with_NeMo.ipynb [hunk 5]
```diff
@@ -0,0 +1,920 @@
...
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "2f142kIQc1Z2",
+        "colab_type": "text"
+      },
+      "source": [
+        "## Transfer learning\n",
+        "\n",
+        "Transfer learning is an important machine learning technique that uses a model’s knowledge of one task to make it perform better on another. Fine-tuning is one of the techniques to perform transfer learning. It is an essential part of the recipe for many state-of-the-art results where a base model is first pretrained on a task with abundant training data and then fine-tuned on different tasks of interest where the training data is less abundant or even scarce.\n",
```
Comment (redoctopus) [inline tutorials/asr/ASR_with_NeMo.ipynb:733]: Add example -- general speech -> domain specific speech, english -> spanish, ...

File: tutorials/asr/ASR_with_NeMo.ipynb [hunk 6]
```diff
@@ -0,0 +1,920 @@
...
+      },
+      "source": [
+        "# To use smaller learning rate\n",
+        "quartznet.setup_optimization(optim_config=DictConfig(new_opt))\n",
+        "# To point to the data we'll use for finetuning\n",
+        "quartznet.setup_training_data(train_data_config=params['model']['train_ds'])\n",
+        "# To point to the data we'll use for finetuning\n",
+        "quartznet.setup_training_data(train_data_config=params['model']['validation_ds'])\n",
+        "trainer = pl.Trainer(gpus=[1], max_epochs=2)\n",
+        "# TODO: OLEKSII FIX NEMO bug\n",
```
Comment (redoctopus) [inline tutorials/asr/ASR_with_NeMo.ipynb:791]: Huh?

File: tutorials/asr/ASR_with_NeMo.ipynb [hunk 7]
```diff
@@ -0,0 +1,920 @@
...
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "2f142kIQc1Z2",
+        "colab_type": "text"
+      },
+      "source": [
+        "## Transfer learning\n",
```
Comment (redoctopus) [inline tutorials/asr/ASR_with_NeMo.ipynb:731]: I think this could go under the "Model Improvements" section.

File: tutorials/asr/ASR_with_NeMo.ipynb [hunk 8]
```diff
@@ -0,0 +1,920 @@
...
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "9glGogaPlR0H",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "source": [
+        "print(quartznet._cfg['spec_augment'])"
```
Comment (redoctopus) [inline tutorials/asr/ASR_with_NeMo.ipynb:825]: Should probably also explain how one would have to add this to the config file to add this into the model.

File: tutorials/asr/ASR_with_NeMo.ipynb [hunk 9]
```diff
@@ -0,0 +1,920 @@
...
+      "metadata": {
+        "id": "wteGqroafWg1",
+        "colab_type": "text"
+      },
+      "source": [
+        "## Under the hood\n",
+        "\n",
+        "NeMo is open-source and we do all our model development in the open. So you can inspect our code.\n",
+        "\n",
+        "In particualr, ``nemo_asr.model.EncDecCTCModel`` is an encoder-decoder model which is construted using several ``Neural Modules`` taken from ``nemo_asr.modules.``. Here is how it's forward pass looks like:\n",
```
Comment (redoctopus) [inline tutorials/asr/ASR_with_NeMo.ipynb:857]: Extra period?

Comment (okuchaiev) [reply tutorials/asr/ASR_with_NeMo.ipynb]: removed

File: tutorials/asr/ASR_with_NeMo.ipynb (hunk 8 above)
Comment (okuchaiev) [reply tutorials/asr/ASR_with_NeMo.ipynb]: added

File: tutorials/asr/ASR_with_NeMo.ipynb (hunk 7 above)
Comment (okuchaiev) [reply tutorials/asr/ASR_with_NeMo.ipynb]: agree - moved

File: tutorials/asr/ASR_with_NeMo.ipynb (hunk 6 above)
Comment (okuchaiev) [reply tutorials/asr/ASR_with_NeMo.ipynb]: fixed, now it actually works :)

File: tutorials/asr/ASR_with_NeMo.ipynb (hunk 5 above)
Comment (okuchaiev) [reply tutorials/asr/ASR_with_NeMo.ipynb:792]: done

File: tutorials/asr/ASR_with_NeMo.ipynb [hunk 10]
```diff
@@ -0,0 +1,920 @@
...
+        "These are followed by two more entries for the blocks that appear at the end of our Jasper model before the CTC loss.\n",
+        "\n",
+        "There are also some entries at the top of the file that specify how we will handle training (`train_ds`) and validation (`validation_ds`) data.\n",
+        "\n",
+        "Using a YAML config such as this is helpful for getting a quick and human-readable overview of what your architecture looks like, and allows you to swap out model and run configurations easily without needing to change your code.\n",
+        "\n",
+        "### Creating Data Manifests\n",
+        "\n",
+        "The first thing we need to do now is to create manifests for our training and evaluation data, which will contain the metadata of our audio files. NeMo data layers take in a standardized manifest format where each line corresponds to one sample of audio, such that the number of lines in a manifest is equal to the number of samples that are represented by that manifest. A line must contain the path to an audio file, the corresponding transcript (or path to a transcript file), and the duration of the audio sample.\n",
```
[the remark this answers is not in the document]
Comment (okuchaiev) [reply tutorials/asr/ASR_with_NeMo.ipynb]: makes sense  - switched them.

File: tutorials/asr/ASR_with_NeMo.ipynb [hunk 11]
```diff
@@ -0,0 +1,920 @@
...
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "ABUDaC5Js7AW",
+        "colab_type": "text"
+      },
+      "source": [
+        "## Training from scratch\n",
+        "\n",
+        "To train from scratch, you need to specify your models architecture\n",
```
[the remark this answers is not in the document]
Comment (okuchaiev) [reply tutorials/asr/ASR_with_NeMo.ipynb]: fixed

File: tutorials/asr/ASR_with_NeMo.ipynb [hunk 12]
```diff
@@ -0,0 +1,920 @@
...
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "ppUm_kuavm_f",
+        "colab_type": "text"
+      },
+      "source": [
+        "That was easy! But there are plenty of scenarious where you would want to fine-tune the model on your own data or even train from scratch. For example, this out-of-the box model will obviously not work for Spanish and would likely perform poorly for telephone audio. So if you have collected your own data, you certainly should attempt to fine-tune or train on it!"
```
[the remark this answers is not in the document]
Comment (okuchaiev) [reply tutorials/asr/ASR_with_NeMo.ipynb]: fixed

File: tutorials/asr/ASR_with_NeMo.ipynb [hunk 13]
```diff
@@ -0,0 +1,920 @@
...
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "KucxoFJhum0i",
+        "colab_type": "text"
+      },
+      "source": [
+        "Next, we'll simply add paths to files we want to transcribe into the list and pass it to our model. Note that it will work for relatively short (<25 seconds) files "
```
[the remark this answers is not in the document]
Comment (okuchaiev) [reply tutorials/asr/ASR_with_NeMo.ipynb]: fixed

File: tutorials/asr/ASR_with_NeMo.ipynb [hunk 14]
```diff
@@ -0,0 +1,920 @@
...
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "v_W8EbYktZE3",
+        "colab_type": "text"
+      },
+      "source": [
+        "## Using out-of-the box model\n",
+        "\n",
+        "NeMo's ASR collection comes with many building blocks and even complete models that we can use for training and evaluation. Moreover, several models come with pre-trained weights. Let's instantiate a complete QuartzNet15x5 model"
```
[the remark this answers is not in the document]
Comment (okuchaiev) [reply tutorials/asr/ASR_with_NeMo.ipynb]: fixed

File: tutorials/asr/ASR_with_NeMo.ipynb [hunk 15]
```diff
@@ -0,0 +1,920 @@
...
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "PXVKBniMlRz5",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "source": [
+        "# --- Config Information ---#\n",
```
[the remark this answers is not in the document]
Comment (okuchaiev) [reply tutorials/asr/ASR_with_NeMo.ipynb:580]: yes

File: tutorials/asr/ASR_with_NeMo.ipynb [hunk 16]
```diff
@@ -0,0 +1,920 @@
...
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "wUmq3p2Aw_5N",
+        "colab_type": "text"
+      },
+      "source": [
+        "NeMo models and modules can be used in any PyTorch code where torch.nn.Module is expected.\n",
```
[the remark this answers is not in the document]
Comment (okuchaiev) [reply tutorials/asr/ASR_with_NeMo.ipynb:601]: added - "Training With PyTorch Lightning"

File: tutorials/asr/ASR_with_NeMo.ipynb (hunk 1 above)
Comment (okuchaiev) [reply tutorials/asr/ASR_with_NeMo.ipynb:668]: added "Monitoring Training"

File: tutorials/asr/ASR_with_NeMo.ipynb (hunk 2 above)
Comment (okuchaiev) [reply tutorials/asr/ASR_with_NeMo.ipynb]: fixed

File: tutorials/asr/ASR_with_NeMo.ipynb [hunk 17]
```diff
@@ -0,0 +1,960 @@
...
+        "\n",
+        "If we open up this config file, we find model section which describes architecture of our model. A model contains an entry labeled `encoder`, with a field called `jasper` that contains a list with multiple entries. Each of the members in this list specifies one block in our model, and looks something like this:\n",
+        "```\n",
+        "- filters: 256\n",
+        "  repeat: 1\n",
+        "  kernel: [11]\n",
+        "  stride: [2]\n",
+        "  dilation: [1]\n",
+        "  dropout: 0.2\n",
+        "  residual: false\n",
```
Comment (redoctopus) [inline tutorials/asr/ASR_with_NeMo.ipynb:561]: I just realized that this is slightly outdated now. Could you update this to add the new fields? (separable, se, se_context_size).

File: tutorials/asr/ASR_with_NeMo.ipynb (hunk 4 above)
Comment (okuchaiev) [reply tutorials/asr/ASR_with_NeMo.ipynb:736]: added inference section
